### ISSUE TITLE
Duplicate holdings on b33039641

### ISSUE THREAD

**Issue by jtweed** [wellcomecollection/platform#5577]
This is very likely related to #5473 and #5492.

We've had a report that an old holdings record is still being displayed on a journal, alongside the correct holdings:

https://wellcomecollection.org/works/jn3mhfbq
https://catalogue.wellcomelibrary.org/record=b3303966~S12

The data is correct in Sierra and on OPAC. There is one holdings (checkin record c11155280) for the bib b33039641.

**Comment by jtweed** [comment]
Another example from Nic that is possibly the same issue:

> Just adding another example here.
> 
> This is a physical journal record that an e-journal checkin got incorrectly attached to. The physical bib b28534542 only has one checkin c11074929 in Sierra (correct) but the checkin for the e-bib c11156466 still remains even after being moved to the correct e-bib b2858191x
> 
> It looks like this online: https://wellcomecollection.org/works/cukjnx6c

**Comment by alexwlchan** [comment]
This looks like an issue with the holdings linker; our stored version of c1115523 has an `updatedDate` of `2022-05-12T21:08:16Z`, but if you retrieve the record from Sierra, it's deleted with a `deletedDate` of `2022-05-12`.

Let me investigate.

**Comment by alexwlchan** [comment]
Hmm, looks like the linker correctly marked it for deletion.

<img width="770" alt="Screenshot 2022-09-01 at 13 42 01" src="https://user-images.githubusercontent.com/301220/187916245-50fb752a-7f7f-4ab6-a9d6-e8f7cfb00738.png">

I'm going to refetch this holdings record through the adapter, and see if I see anything interesting in the logs.

**Comment by alexwlchan** [comment]
Aha, so there's an issue that the updatedDate on the modified and deleted records are identical, and the merger wouldn't remove the record if that was the case.

I've deployed a patched merger, which should fix this if I can get the holdings reader to re-fetch the record – but I am completely unable to do so, grr.

**Comment by alexwlchan** [comment]
I wrote a quick script using [my Python Sierra API client](https://github.com/wellcomecollection/sierra_api) to get a list of all deleted holdings IDs:

<details>
<summary>get_deleted_holdings.py</summary>

```python
import os
import sys

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from api_client import catalogue_client


if __name__ == "__main__":
    client = catalogue_client()

    for holdings in client.get_objects('/holdings', params={'deleted': True}):
        print(holdings['id'])
```

</details>

That spat out 75k IDs: [deleted_holdings.txt](https://github.com/wellcomecollection/platform/files/9470410/deleted_holdings.txt)

Then I wrote a script that queries the Sierra linker table to find all the bib IDs which might be linked to now-deleted holdings records:

<details><summary>find_affected_bibs.py</summary>

```python
#!/usr/bin/env python3

import sys
import boto3, tqdm


def scan_table(sess, *, TableName, **kwargs):
    """
    Generates all the items in a DynamoDB table.

    :param dynamo_client: A boto3 client for DynamoDB.
    :param TableName: The name of the table to scan.

    Other keyword arguments will be passed directly to the Scan operation.
    See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb.html#DynamoDB.Client.scan

    See https://alexwlchan.net/2020/05/getting-every-item-from-a-dynamodb-table-with-python/

    """
    dynamo_client = sess.resource("dynamodb").meta.client

    paginator = dynamo_client.get_paginator("scan")

    for page in paginator.paginate(TableName=TableName, **kwargs):
        yield from page["Items"]


if __name__ == '__main__':
    sess = boto3.Session()

    deleted_holdings_ids = set(line.strip() for line in open('deleted_holdings.txt'))

    affected_bib_ids = set()

    for item in tqdm.tqdm(scan_table(sess, TableName="sourcedata-sierra-holdings-20200604"), total=73351):
        if item['id'] not in deleted_holdings_ids:
            continue

        # This holdings record has never been linked to a bib in the
        # history of the pipeline
        if not item['payload']['bibIds'] and not item['payload']['unlinkedBibIds']:
            continue

        if item['payload']['bibIds']:
            print('!!!', f'deleted holdings record {item["id"]} has bibIds', item['payload'], file=sys.stderr)

        for bib_id in item['payload']['unlinkedBibIds']:
            affected_bib_ids.add(bib_id)

    with open('affected_bib_ids.txt', 'w') as outfile:
        outfile.write('\n'.join(sorted(affected_bib_ids)))
```

</details>

That spat out ~3.5k bib IDs: [affected_bib_ids.txt](https://github.com/wellcomecollection/platform/files/9470436/affected_bib_ids.txt)

It also suggests we've failed to pick up the deletion of approximately 20 holdings records:

```
!!! deleted holdings record 1089032 has bibIds {'modifiedDate': Decimal('1444660272000'), 'bibIds': ['2527751'], 'unlinkedBibIds': []}
!!! deleted holdings record 1089150 has bibIds {'modifiedDate': Decimal('1444742766000'), 'bibIds': ['2508760'], 'unlinkedBibIds': []}
!!! deleted holdings record 1089212 has bibIds {'modifiedDate': Decimal('1445264271000'), 'bibIds': ['2721274'], 'unlinkedBibIds': []}
!!! deleted holdings record 1089214 has bibIds {'modifiedDate': Decimal('1445264277000'), 'bibIds': ['2756123'], 'unlinkedBibIds': []}
!!! deleted holdings record 1092428 has bibIds {'modifiedDate': Decimal('1570806540000'), 'bibIds': ['2530982'], 'unlinkedBibIds': []}
!!! deleted holdings record 1092748 has bibIds {'modifiedDate': Decimal('1455538581000'), 'bibIds': ['2509985'], 'unlinkedBibIds': []}
!!! deleted holdings record 1095000 has bibIds {'modifiedDate': Decimal('1460547004000'), 'bibIds': ['2502465'], 'unlinkedBibIds': []}
!!! deleted holdings record 1095461 has bibIds {'modifiedDate': Decimal('1583003042000'), 'bibIds': ['2506781'], 'unlinkedBibIds': []}
!!! deleted holdings record 1095601 has bibIds {'modifiedDate': Decimal('1583003044000'), 'bibIds': ['2512722'], 'unlinkedBibIds': []}
!!! deleted holdings record 1095676 has bibIds {'modifiedDate': Decimal('1583003045000'), 'bibIds': ['2532048'], 'unlinkedBibIds': []}
!!! deleted holdings record 1100022 has bibIds {'modifiedDate': Decimal('1463481691000'), 'bibIds': ['2819455'], 'unlinkedBibIds': []}
!!! deleted holdings record 1107609 has bibIds {'modifiedDate': Decimal('1495463408000'), 'bibIds': ['2855301'], 'unlinkedBibIds': []}
!!! deleted holdings record 1109521 has bibIds {'modifiedDate': Decimal('1587750466000'), 'bibIds': ['2509985'], 'unlinkedBibIds': []}
!!! deleted holdings record 1109542 has bibIds {'modifiedDate': Decimal('1587750470000'), 'bibIds': ['2519033'], 'unlinkedBibIds': []}
!!! deleted holdings record 1110175 has bibIds {'modifiedDate': Decimal('1582036458000'), 'bibIds': ['2819455'], 'unlinkedBibIds': []}
!!! deleted holdings record 1110348 has bibIds {'modifiedDate': Decimal('1582037546000'), 'bibIds': ['2502680'], 'unlinkedBibIds': []}
!!! deleted holdings record 1110730 has bibIds {'modifiedDate': Decimal('1582037596000'), 'bibIds': ['2509985'], 'unlinkedBibIds': []}
!!! deleted holdings record 1111180 has bibIds {'modifiedDate': Decimal('1583089921000'), 'bibIds': ['2509266'], 'unlinkedBibIds': []}
!!! deleted holdings record 1111191 has bibIds {'modifiedDate': Decimal('1582037819000'), 'bibIds': ['2506781'], 'unlinkedBibIds': []}
!!! deleted holdings record 1111253 has bibIds {'modifiedDate': Decimal('1582037826000'), 'bibIds': ['2501174'], 'unlinkedBibIds': []}
!!! deleted holdings record 1111459 has bibIds {'modifiedDate': Decimal('1582037850000'), 'bibIds': ['2512722'], 'unlinkedBibIds': []}
```

**Comment by alexwlchan** [comment]
Querying the first holdings record on that list is revealing:

```console
$ python3 scripts/get_catalogue_record.py c1089032
{
  "entries": [
    {
      "bibIds": [],
      "deleted": true,
      "deletedDate": "2022-08-16",
      "fixedFields": {
        "80": {
          "label": "REC TYPE",
          "value": "c"
        },
        "81": {
          "label": "RECORD #",
          "value": "1089032"
        },
        "84": {
          "label": "UPDATED",
          "value": "2022-08-15T23:00:00Z"
        },
        "85": {
          "label": "REVISIONS",
          "value": "1"
        },
        "86": {
          "label": "AGENCY",
          "value": "1"
        },
        "98": {
          "label": "PDATE",
          "value": "2015-10-12T14:31:12Z"
        }
      },
      "id": 1089032,
      "itemIds": [],
      "updatedDate": "2015-10-12T14:31:12Z",
      "varFields": []
    }
  ],
  "start": 0,
  "total": 1
}
```

The record was deleted on `2022-08-16`, but the `updatedDate` is `2015-10-12T14:31:12Z`. Unhelpful!

**Comment by alexwlchan** [comment]
There are ~73k holdings records in the pipeline, so I've sent a new batch of windows to the reader to re-fetch all the holdings from Sierra. I hope this will flush out the issues automatically without me having to do data surgery. 🤞 

It doesn't give us a permanent fix, but it will at least address the current issues.

**Comment by alexwlchan** [comment]
Reloading the entire pipeline has fixed part of the issue – in particular, all the records it failed to delete. But it _still_ hasn't fixed the original thing that was reported, which is incredibly frustrating.

I think I might have to manually patch these two records.

**Comment by alexwlchan** [comment]
Looking at Nic’s second example, there are three bib records linked to b2853454:

* c1107492, which is the thing we want
* c1115496, which has been updated in Sierra but we seemingly haven't picked up that change
* c1115670, which looks like an online location we don't want

**Comment by alexwlchan** [comment]
Looking at https://wellcomecollection.org/works/jn3mhfbq, this actually links to b330396**4**, not b330396**6** (which does have the correct single holdings record).

**Comment by alexwlchan** [comment]
tl;dr: there's an issue in how we fetch holdings records from Sierra which means some updates are being lost; I don't have a better way to fix this than reloading all the records from Sierra.

There's an outstanding issue with b2853454, but I think that might be a Sierra data issue – I've asked Nic about it in Slack.

Once that's answered, I think this is fixed as best we can do.

**Comment by alexwlchan** [comment]
It looks like both the problem records are now fixed, plus as many other records with similar problems that I could find. Closing as done.
